core/src/main/scala/org/ensime/indexer/graph/GraphService.scala

```diff
@@ -0,0 +1,368 @@
+    case fn: FieldName => Some(Field(name.fqnString, None, None, None, null, None))
+    case mn: MethodName => Some(Method(name.fqnString, None, None, null, None))
+    case _ => None
+  }
+}
+
+sealed trait Hierarchy
+object Hierarchy {
+  sealed trait Direction
+  case object Supertypes extends Direction
+  case object Subtypes extends Direction
+}
+
+final case class TypeHierarchy(aClass: ClassDef, typeRefs: Seq[Hierarchy]) extends Hierarchy
+
+final case class ClassDef(
+    fqn: String,
+    file: String, // the underlying file
+    path: String, // the VFS handle (e.g. classes in jars)
+    source: Option[String],
+    line: Option[Int],
+    access: Access,
+    scalaName: Option[String],
+    scalapDeclaredAs: Option[DeclaredAs]
+) extends FqnSymbol with Hierarchy {
+  override def declAs: DeclaredAs = scalapDeclaredAs.getOrElse(DeclaredAs.Class)
+}
+
+sealed trait Member extends FqnSymbol
+
+final case class Field(
+    fqn: String,
+    internal: Option[String],
+    line: Option[Int],
+    source: Option[String],
+    access: Access,
+    scalaName: Option[String]
+) extends Member {
+  override def declAs: DeclaredAs = DeclaredAs.Field
+}
+
+final case class Method(
+    fqn: String,
+    line: Option[Int],
+    source: Option[String],
+    access: Access,
+    scalaName: Option[String]
+) extends Member {
+
+  override def declAs: DeclaredAs = DeclaredAs.Method
+}
+
+final case class FileCheck(filename: String, timestamp: Timestamp) {
+  def file(implicit vfs: EnsimeVFS): FileObject = vfs.vfile(filename)
+  def lastModified: Long = timestamp.getTime
+  def changed(implicit vfs: EnsimeVFS): Boolean = file.getContent.getLastModifiedTime != lastModified
+}
+object FileCheck extends ((String, Timestamp) => FileCheck) {
+  def apply(f: FileObject): FileCheck = {
+    val name = f.getName.getURI
+    val ts = if (f.exists()) new Timestamp(f.getContent.getLastModifiedTime)
+    else new Timestamp(-1L)
+    FileCheck(name, ts)
+  }
+
+  def fromPath(path: String)(implicit vfs: EnsimeVFS): FileCheck = apply(vfs.vfile(path))
+}
+
+// core/it:test-only *Search* -- -z prestine
+class GraphService(dir: File) extends SLF4JLogging {
+  import org.ensime.indexer.graph.GraphService._
+
+  implicit val FqnSymbolS: BigDataFormat[FqnSymbol] = cachedImplicit
+  implicit val MemberS: BigDataFormat[Member] = cachedImplicit
+  implicit val FileCheckS: BigDataFormat[FileCheck] = cachedImplicit
+  implicit val ClassDefS: BigDataFormat[ClassDef] = cachedImplicit
+  implicit val MethodS: BigDataFormat[Method] = cachedImplicit
+  implicit val FieldS: BigDataFormat[Field] = cachedImplicit
+
+  import shapeless._
+  implicit val UniqueFileCheckV = LensId("filename", lens[FileCheck] >> 'filename)
+  implicit val FieldV = LensId("fqn", lens[Field] >> 'fqn)
+  implicit val MethodV = LensId("fqn", lens[Method] >> 'fqn)
+  implicit val ClassDefV = LensId("fqn", lens[ClassDef] >> 'fqn)
+
+  private implicit val FqnSymbolLens = new Lens[FqnSymbol, String] {
```

Comment:
perhaps using `Lens` here in the first place was heavy handed, since we only want to use it for reading values (not returning copies with a value changed).

It would be a lot simpler if you just defined

```scala
  private implicit val UniqueFqnSymbolV: BigDataFormatId[FqnSymbol, String] =
    new BigDataFormatId[FqnSymbol, String] {
      override def key = "fqn"
      override def value(t: FqnSymbol): String = t.fqn
    }
```

without bothering with `Lens`.